Json forget

d3zd3z opened the pull request:
<!--
Thank you very much for contributing code or documentation to restic! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

What is the purpose of this change? What does it change?
--------------------------------------------------------
This makes the `--json` option work with the `forget` command.

<!--
Describe the changes here, as detailed as needed.
-->

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
I created issue #2184 to describe this.

Closes #2184

<!--
Link issues and relevant forum posts here.
-->

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
  (it could be argued that this is already described in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

d3zd3z committed:
Add `--json` support to `forget` command

Fixes #2184

d3zd3z committed:
Add test for `--json` support for `forget` command

This adds a test of the json output of the forget command, by running it
once, asking it to keep one snapshot, and verifying that the output has
the right number of snapshots listed in the Keep and Remove fields of
the result.

d3zd3z committed:
Add changelog for #2184

fd0 commented on changelog/unreleased/issue-2184:
```diff
@@ -0,0 +1,12 @@
+information about what is (or would-be) kept and removed from the
+repository.
+
+The text in the paragraphs is written in past tense. The last section is a list
```
Remove this paragraph please ;)

mholt replied:
Gee I dunno, I kinda like it! Reads like a sudden plot twist to a self-describing existential crisis. 😜

d3zd3z replied:
Removed.

fd0 reviewed (changes requested):
Looks great already, thanks for your contribution! It'd be nice if you could also integrate the reasons each snapshot is kept (which is part of the normal output, but not included in the JSON output yet). Would you like to do that in this PR? Or open a new PR later on?

d3zd3z committed:
Include reasons in json output of forget

This dumps the reasons as well as the list of keeps and removes with the
output from the forget command.

d3zd3z commented:
This includes the reasons, although this is kind of simplistic in that it just dumps the reasons out. Does it seem useful this way, or should I try to correlate each reason with each keep and remove?

fd0 commented:
Awesome, thanks! looks good!
